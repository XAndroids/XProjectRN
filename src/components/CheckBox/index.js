import React from 'react'
import {TouchableOpacity, Image} from 'react-native'

import PropTypes from "prop-types";

//函数式组件:XCheckBox
//参考：https://github.com/XLibrarySources/react-native-elements/blob/next/src/checkbox/CheckBox.js
const CheckBox = (props) => {
    const icon = props.isChecked ? props.checkedIcon : props.uncheckedIcon;

    return (
        <TouchableOpacity onPress={props.onCheckedChange}>
            <Image testID={'checkbox_image'} style={[props.style]} source={icon}/>
        </TouchableOpacity>
    );
};

CheckBox.propTypes = {
    isChecked: PropTypes.bool.isRequired,//没有默认值，都声明呈required
    checkedIcon: PropTypes.object.isRequired,//FIXME 这是用PropTypes.object检测是否合适？？
    uncheckedIcon: PropTypes.object.isRequired,
    onCheckedChange: PropTypes.func
};

export default CheckBox
