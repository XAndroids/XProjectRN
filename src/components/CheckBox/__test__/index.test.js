//不使用箭头函数，使用funtion传递参数
import React from "react";
import {fireEvent, render, shallow} from "react-native-testing-library";
import {Image} from 'react-native'
import CheckBox from "../index";

describe('CheckBox property test', function () {
    it('property isChecked/checkedIcon/uncheckedIcon test ', function () {
        const {getByName} = render(<CheckBox isChecked={false}
                                             checkedIcon={require('../../../../img/publisherlist_checkbox_selected.png')}
                                             uncheckedIcon={require('../../../../img/publisherlist_checkbox_add.png')}/>);
        //getByName：通过React Native组件的名称查找
        expect(getByName('Image').props.source.testUri).toBe('../../../img/publisherlist_checkbox_add.png');

        const {getByType} = render(<CheckBox isChecked={true}
                                             checkedIcon={require('../../../../img/publisherlist_checkbox_selected.png')}
                                             uncheckedIcon={require('../../../../img/publisherlist_checkbox_add.png')}/>);
        //getByType：通过React Native组件的类型查找
        expect(getByType(Image).props.source.testUri).toBe('../../../img/publisherlist_checkbox_selected.png');
    });

    it('property style test ', function () {
        const {getByTestId} = render(<CheckBox style={{width: 50, height: 50, marginRight: 6}}
                                               isChecked={false}
                                               checkedIcon={require('../../../../img/publisherlist_checkbox_selected.png')}
                                               uncheckedIcon={require('../../../../img/publisherlist_checkbox_add.png')}/>);
        const style = getByTestId('checkbox_image').props.style;
        expect(style).toEqual([{width: 50, height: 50, marginRight: 6}]);
    });
});

describe('CheckBox fun test', () => {
    it('fun onCheckedChange test ', function () {
        const onCheckedChange = jest.fn();
        const {getByTestId} = render(<CheckBox isChecked={true}
                                               checkedIcon={require('../../../../img/publisherlist_checkbox_selected.png')}
                                               uncheckedIcon={require('../../../../img/publisherlist_checkbox_add.png')}
                                               onCheckedChange={onCheckedChange}/>);
        //通过子组件，调用父组件的onPress()方法
        fireEvent(getByTestId('checkbox_image'), 'onPress');
        expect(onCheckedChange).toHaveBeenCalled();
    });
});

describe('CheckBox Snapshot test', () => {
    test('Snapshot test', () => {
        const onCheckedChange = jest.fn();
        const {toJSON} = render(<CheckBox isChecked={true}
                                          checkedIcon={require('../../../../img/publisherlist_checkbox_selected.png')}
                                          uncheckedIcon={require('../../../../img/publisherlist_checkbox_add.png')}
                                          onCheckedChange={onCheckedChange}/>);
        expect(toJSON()).toMatchSnapshot();
    });
});

describe('CheckBox Shallow test', () => {
    test('Shallow test', () => {
        const onCheckedChange = jest.fn();
        const {output} = shallow(<CheckBox isChecked={true}
                                           checkedIcon={require('../../../../img/publisherlist_checkbox_selected.png')}
                                           uncheckedIcon={require('../../../../img/publisherlist_checkbox_add.png')}
                                           onCheckedChange={onCheckedChange}/>);
        expect(output).toMatchSnapshot();
    });
});
