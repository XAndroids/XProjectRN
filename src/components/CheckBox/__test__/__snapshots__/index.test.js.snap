// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox Shallow test Shallow test 1`] = `
<TouchableOpacity
  activeOpacity={0.2}
  onPress={[MockFunction]}
>
  <Image
    source={
      Object {
        "testUri": "../../../img/publisherlist_checkbox_selected.png",
      }
    }
    style={
      Array [
        undefined,
      ]
    }
    testID="checkbox_image"
  />
</TouchableOpacity>
`;

exports[`CheckBox Snapshot test Snapshot test 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../img/publisherlist_checkbox_selected.png",
      }
    }
    style={
      Array [
        undefined,
      ]
    }
    testID="checkbox_image"
  />
</View>
`;
